nixos/modules/hardware/printers.nix

```diff
@@ -0,0 +1,111 @@
+        '';
+      };
+      ensurePrinters = mkOption {
+        description = ''
+          Will regularly ensure that the given CUPS printers are configured as declared here.
+          If a printer's options are manually changed afterwards, they will be overwritten eventually.
+          This option will never delete any printer, even if removed from this list.
+          You can check existing printers with <literal>lpstat -s</literal>
+          and remove printers with <literal>lpadmin -x \<printer-name\></literal>.
+          Printers not listed here can still be manually configured.
+        '';
+        type = types.listOf (types.submodule {
+          options = {
+            name = mkOption {
+              type = printerName;
+              example = "BrotherHL_Workroom";
+              description = ''
+                Name of the printer / printer queue.
+                May contain any printable characters except "/", "#", and space.
+              '';
+            };
+            location = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "Workroom";
+              description = ''
+                Optional human-readable location.
+              '';
+            };
+            description = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "Brother HL-5140";
+              description = ''
+                Optional human-readable description.
+              '';
+            };
+            deviceUri = mkOption {
+              type = types.str;
+              example = [
+                "ipp://printserver.local/printers/BrotherHL_Workroom"
+                "usb://HP/DESKJET%20940C?serial=CN16E6C364BH"
+              ];
+              description = ''
+                How to reach the printer.
+                <literal>lpinfo -v</literal> shows a list of supported device URIs and schemes.
+              '';
+            };
+            model = mkOption {
+              type = types.str;
+              example = "gutenprint.5.2://brother-hl-5140/expert";
+              description = ''
+                Location of the ppd driver file for the printer.
+                <literal>lpinfo -m</literal> shows a list of supported models.
+              '';
+            };
+          };
+        });
+        default = [];
+      };
+    };
+  };
+
+  config = mkIf (cfg.ensurePrinters != [] && config.services.printing.enable) {
+    systemd.services."ensure-printers" = let
+      cupsUnit = if config.services.printing.startWhenNeeded then "cups.socket" else "cups.service";
+    in {
+      description = "Ensure NixOS-configured CUPS printers";
+      wantedBy = [ "multi-user.target" ];
+      requires = [ cupsUnit ];
+      # for the service, this is actually not enough, because the socket file might still not exist at that point
+      after = [ cupsUnit ];
```

Comment:
Can you expand on this? I don't get the problem. Why does "the socket file might still not exist" matter when it's a service?

Reply:
Tried to elaborate. The gist is: regardless of being socket activated or an always running service, commands like `lpinfo -v` interact with cups through a UNIX socket. With service-based cups, this socket file is not yet created when the cups service gets into “activated” stage, which makes commands fail, while with socket activation, systemd will create the socket file upfront.